// @ff/core/Document.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2019 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const Publisher_1 = require("./Publisher");
const uniqueId_1 = require("./uniqueId");
class Document extends Publisher_1.default {
    constructor(json) {
        super();
        this.addEvents("update", "dispose");
        if (json) {
            this.fromJSON(json);
        }
        else {
            this._data = this.init();
        }
    }
    static generateId() {
        return uniqueId_1.default();
    }
    get id() {
        return this._data["id"];
    }
    get data() {
        return this._data;
    }
    set(key, value) {
        this._data[key] = value;
        this.update();
    }
    get(key) {
        return this._data[key];
    }
    update() {
        this.emit({ type: "update", document: this });
    }
    dispose() {
        this.emit({ type: "dispose", document: this });
        this._data = null;
    }
    fromJSON(json) {
        this._data = {};
        this.inflate(json, this._data);
        this.update();
        return this;
    }
    toJSON(json) {
        json = json || {};
        this.deflate(this._data, json);
        return json;
    }
    init() {
        return {};
    }
    inflate(json, data) {
        Object.assign(data, json);
    }
    deflate(data, json) {
        json = json || {};
        Object.assign(json, this._data);
    }
}
exports.default = Document;
