// @ff/core/Vector3.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2019 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const Vector2_1 = require("./Vector2");
/**
 * 3-dimensional vector.
 */
class Vector3 {
    /**
     * Constructs a new vector with the given x, y, and z values.
     * Omitted or invalid values are set to zero.
     * @param x
     * @param y
     * @param z
     */
    constructor(x, y, z) {
        this.x = x || 0;
        this.y = y || 0;
        this.z = z || 0;
    }
    /**
     * Returns a new vector with all components set to zero.
     */
    static makeZeros() {
        return new Vector3(0, 0, 0);
    }
    /**
     * Returns a new vector with all components set to one.
     */
    static makeOnes() {
        return new Vector3(1, 1, 1);
    }
    /**
     * Returns a new unit-length vector, parallel to the X axis.
     */
    static makeUnitX() {
        return new Vector3(1, 0, 0);
    }
    /**
     * Returns a new unit-length vector, parallel to the Y axis.
     */
    static makeUnitY() {
        return new Vector3(0, 1, 0);
    }
    /**
     * Returns a new unit-length vector, parallel to the Z axis.
     */
    static makeUnitZ() {
        return new Vector3(0, 0, 1);
    }
    /**
     * Returns a new vector with components set from the given vector.
     * @param vector
     */
    static makeCopy(vector) {
        return new Vector3(vector.x, vector.y, vector.z);
    }
    /**
     * Returns a new vector with each component set to the given scalar value.
     * @param scalar
     */
    static makeFromScalar(scalar) {
        return new Vector3(scalar, scalar, scalar);
    }
    /**
     * Returns a new vector with components set from the values of the given array.
     * @param array
     */
    static makeFromArray(array) {
        return new Vector3(array[0], array[1], array[2]);
    }
    /**
     * Returns a string representation of the given vector.
     * @param vector
     */
    static toString(vector) {
        return `[${vector.x}, ${vector.y}, ${vector.z}]`;
    }
    /**
     * Copies the components of the given vector to this.
     * @param vector
     */
    copy(vector) {
        this.x = vector.x;
        this.y = vector.y;
        this.z = vector.z;
        return this;
    }
    /**
     * Sets the components to the given values.
     * @param x
     * @param y
     * @param z
     */
    set(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
        return this;
    }
    /**
     * Sets each component to the given scalar value.
     * @param scalar
     */
    setFromScalar(scalar) {
        this.x = scalar;
        this.y = scalar;
        this.z = scalar;
        return this;
    }
    /**
     * Sets the components to the values of the given array.
     * @param array
     * @param offset Optional start index of the array. Default is 0.
     */
    setFromArray(array, offset = 0) {
        this.x = array[offset];
        this.y = array[offset + 1];
        this.z = array[offset + 2];
        return this;
    }
    /**
     * Sets all components to zero.
     */
    setZeros() {
        this.x = 0;
        this.y = 0;
        this.z = 0;
        return this;
    }
    /**
     * Sets all components to one.
     */
    setOnes() {
        this.x = 1;
        this.y = 1;
        this.z = 1;
        return this;
    }
    /**
     * Makes this a unit vector parallel to the X axis.
     */
    setUnitX() {
        this.x = 1;
        this.y = 0;
        this.z = 0;
        return this;
    }
    /**
     * Makes this a unit vector parallel to the Y axis.
     */
    setUnitY() {
        this.x = 0;
        this.y = 1;
        this.z = 0;
        return this;
    }
    /**
     * Makes this a unit vector parallel to the Z axis.
     */
    setUnitZ() {
        this.x = 0;
        this.y = 0;
        this.z = 1;
        return this;
    }
    /**
     * Adds the given vector to this.
     * @param other
     */
    add(other) {
        this.x += other.x;
        this.y += other.y;
        this.z += other.z;
        return this;
    }
    /**
     * Subtracts the given vector from this.
     * @param other
     */
    sub(other) {
        this.x -= other.x;
        this.y -= other.y;
        this.z -= other.z;
        return this;
    }
    /**
     * Multiplies each component with the corresponding component of the given vector.
     * @param other
     */
    mul(other) {
        this.x *= other.x;
        this.y *= other.y;
        this.z *= other.z;
        return this;
    }
    /**
     * Divides each component by the corresponding component of the given vector.
     * @param other
     */
    div(other) {
        this.x /= other.x;
        this.y /= other.y;
        this.z /= other.z;
        return this;
    }
    /**
     * Adds the given scalar to each component.
     * @param scalar
     */
    addScalar(scalar) {
        this.x += scalar;
        this.y += scalar;
        this.z += scalar;
        return this;
    }
    /**
     * Subtracts the given scalar from each component.
     * @param scalar
     */
    subScalar(scalar) {
        this.x -= scalar;
        this.y -= scalar;
        this.z -= scalar;
        return this;
    }
    /**
     * Multiplies each component with the given scalar.
     * @param scalar
     */
    mulScalar(scalar) {
        this.x *= scalar;
        this.y *= scalar;
        this.z *= scalar;
        return this;
    }
    /**
     * Divides each component by the given scalar.
     * @param scalar
     */
    divScalar(scalar) {
        this.x /= scalar;
        this.y /= scalar;
        this.z /= scalar;
        return this;
    }
    /**
     * Translates the vector by the given offsets.
     * @param tx
     * @param ty
     * @param tz
     */
    translate(tx, ty, tz) {
        this.x += tx;
        this.y += ty;
        this.z += tz;
        return this;
    }
    /**
     * Rotates the vector by the given angle about the x-axis.
     * @param angle rotation angle in radians.
     */
    rotateX(angle) {
        const co = Math.cos(angle);
        const si = Math.sin(angle);
        const y = this.y, z = this.z;
        this.y = co * y - si * z;
        this.z = si * y + co * z;
        return this;
    }
    /**
     * Rotates the vector by the given angle about the y-axis.
     * @param angle rotation angle in radians.
     */
    rotateY(angle) {
        const co = Math.cos(angle);
        const si = Math.sin(angle);
        const x = this.x, z = this.z;
        this.x = co * x + si * z;
        this.z = -si * x + co * z;
        return this;
    }
    /**
     * Rotates the vector by the given angle about the z-axis.
     * @param angle rotation angle in radians.
     */
    rotateZ(angle) {
        const co = Math.cos(angle);
        const si = Math.sin(angle);
        const x = this.x, y = this.y;
        this.x = co * x - si * y;
        this.y = si * x + co * y;
        return this;
    }
    /**
     * Scales the vector by the given factors.
     * @param sx
     * @param sy
     * @param sz
     */
    scale(sx, sy, sz) {
        this.x *= sx;
        this.y *= sy;
        this.z *= sz;
        return this;
    }
    /**
     * Inverts each component of this, e.g. x = 1 / x, ...
     */
    invert() {
        this.x = 1 / this.x;
        this.y = 1 / this.y;
        this.z = 1 / this.z;
        return this;
    }
    /**
     * Negates each component of this, e.g. x = -x, ...
     */
    negate() {
        this.x = -this.x;
        this.y = -this.y;
        this.z = -this.z;
        return this;
    }
    /**
     * Normalizes the vector, making it a unit vector.
     */
    normalize() {
        const f = 1 / Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
        this.x *= f;
        this.y *= f;
        this.z *= f;
        return this;
    }
    /**
     * Makes this vector homogeneous by dividing all components by the z component.
     */
    homogenize() {
        this.x /= this.z;
        this.y /= this.z;
        this.z = 1;
        return this;
    }
    /**
     * Returns the dot product of this and the given vector.
     * @param other
     */
    dot(other) {
        return this.x * other.x + this.y * other.y + this.z * other.z;
    }
    /**
     * Assigns to this the cross product between this and the given vector.
     * @param other
     */
    cross(other) {
        const x = this.y * other.z - this.z * other.y;
        const y = this.z * other.x - this.x * other.z;
        const z = this.x * other.y - this.y * other.x;
        this.x = x;
        this.y = y;
        this.z = z;
        return this;
    }
    /**
     * Returns the 2-norm (length) of this.
     */
    length() {
        const x = this.x, y = this.y, z = this.z;
        return Math.sqrt(x * x + y * y + z * z);
    }
    /**
     * Returns the squared 2-norm of this, i.e. the dot product of the vector with itself.
     */
    lengthSquared() {
        const x = this.x, y = this.y, z = this.z;
        return x * x + y * y + z * z;
    }
    /**
     * Returns the distance between this and other.
     * @param other
     */
    distanceTo(other) {
        const dx = other.x - this.x;
        const dy = other.y - this.y;
        const dz = other.z - this.z;
        return Math.sqrt(dx * dx + dy * dy + dz * dz);
    }
    /**
     * Returns the smallest component.
     */
    min() {
        return this.x < this.y ? (this.x < this.z ? this.x : this.z) : (this.y < this.z ? this.y : this.z);
    }
    /**
     * Returns the largest component.
     */
    max() {
        return this.x > this.y ? (this.x > this.z ? this.x : this.z) : (this.y > this.z ? this.y : this.z);
    }
    /**
     * Returns true if all components are exactly zero.
     */
    isZero() {
        return this.x === 0 && this.y === 0 && this.z === 0;
    }
    /**
     * Returns a clone.
     */
    clone() {
        return new Vector3(this.x, this.y, this.z);
    }
    /**
     * Returns an array with the components of this.
     * @param array Optional destination array.
     * @param offset Optional start index of the array. Default is 0.
     */
    toArray(array, offset) {
        if (array) {
            if (offset === undefined) {
                offset = 0;
            }
            array[offset] = this.x;
            array[offset + 1] = this.y;
            array[offset + 2] = this.z;
            return array;
        }
        return [
            this.x,
            this.y,
            this.z
        ];
    }
    /**
     * Returns a [[Vector2]] with the x and y components of this.
     * @param vector Optional destination vector.
     */
    toVector2(vector) {
        if (vector) {
            vector.x = this.x;
            vector.y = this.y;
            return vector;
        }
        return new Vector2_1.default(this.x, this.y);
    }
    /**
     * Returns a text representation.
     */
    toString() {
        return Vector3.toString(this);
    }
}
exports.default = Vector3;
Vector3.zeros = new Vector3(0, 0, 0);
Vector3.ones = new Vector3(1, 1, 1);
Vector3.unitX = new Vector3(1, 0, 0);
Vector3.unitY = new Vector3(0, 1, 0);
Vector3.unitZ = new Vector3(0, 0, 1);
