// @ff/react/DockSprite.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
class DockSprite extends React.Component {
    constructor(props) {
        super(props);
    }
    render() {
        const { className, dockArea } = this.props;
        let style = Object.assign({}, DockSprite.style);
        switch (dockArea) {
            case "none":
                return null;
            case "left":
                style.right = "50%";
                break;
            case "right":
                style.left = "50%";
                break;
            case "top":
                style.bottom = "50%";
                break;
            case "bottom":
                style.top = "50%";
                break;
        }
        return (React.createElement("div", { className: className, style: style }));
    }
}
exports.default = DockSprite;
DockSprite.defaultProps = {
    className: "ff-dock-sprite",
};
DockSprite.style = {
    boxSizing: "border-box",
    position: "absolute",
    top: "0", left: "0", right: "0", bottom: "0",
    pointerEvents: "none"
};
