// @ff/react/DockTabContainer.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
const TabContainer_1 = require("./TabContainer");
const DockTabContent_1 = require("./DockTabContent");
class DockTabContainer extends TabContainer_1.default {
    constructor(props) {
        super(props);
        this.onSelect = this.onSelect.bind(this);
        this.onDockableDrop = this.onDockableDrop.bind(this);
    }
    renderContentContainer(children) {
        return (React.createElement(DockTabContent_1.DockTabContentContainer, { className: "ff-content", onDrop: this.onDockableDrop }, children));
    }
    onSelect(event) {
        const { id, index, onTabSelect } = this.props;
        if (onTabSelect) {
            onTabSelect({
                tabId: event.id,
                tabIndex: event.index,
                id,
                index,
                sender: this
            });
        }
    }
    onDockableDrop(event) {
        const { id, index, onTabDrop } = this.props;
        if (onTabDrop) {
            onTabDrop({
                sourceTabId: event.sourceTabId,
                sourceTabIndex: event.sourceTabIndex,
                tabId: this.state.activeTabId,
                tabIndex: this.state.activeTabIndex,
                dockArea: event.dockArea,
                id,
                index,
                sender: this
            });
        }
    }
}
exports.default = DockTabContainer;
DockTabContainer.defaultProps = {
    className: "ff-dock-tab-container"
};
