// @ff/react/Dropdown.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
class Dropdown extends React.Component {
    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
    }
    render() {
        const { className, style, options } = this.props;
        const optionElements = options.map((option, index) => (React.createElement("option", { key: option + index, label: option }, option)));
        const styles = Object.assign({}, Dropdown.style, style);
        const touchActionProp = { "touch-action": "none" };
        return (React.createElement("div", { className: className },
            React.createElement("select", Object.assign({ style: styles, tabIndex: 0, onChange: this.onChange }, touchActionProp), optionElements)));
    }
    onChange(e) {
        const { id, index, onSelect } = this.props;
        if (onSelect) {
            const select = e.target;
            const selectedIndex = select.selectedIndex;
            const option = select[index];
            onSelect({ selectedIndex, option: option.label, id, index, sender: this });
        }
    }
}
exports.default = Dropdown;
Dropdown.defaultProps = {
    className: "ff-control ff-dropdown",
    options: []
};
Dropdown.style = {
    touchAction: "none",
    cursor: "pointer"
};
