// @ff/react/FlexItem.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
/**
 * Wraps its children in a div element. Flex layout settings grow, shrink
 * and basis are exposed as component properties. See also [[FlexContainer]].
 * This is a React stateless functional component.
 * @param {IFlexItemProps & {children?: React.ReactNode}} props
 * @returns {React.SFC<IFlexItemProps>}
 * @constructor
 */
const FlexItem = function (props) {
    const { className, style, elementRef, grow, shrink, basis, children } = props;
    const defaultStyle = {
        boxSizing: "border-box",
        position: "relative",
        flex: `${grow} ${shrink} ${basis}`
    };
    const styles = Object.assign({}, defaultStyle, style);
    return (React.createElement("div", { className: className, style: styles, ref: elementRef }, children));
};
FlexItem.defaultProps = {
    className: "ff-flex-item",
    grow: 1,
    shrink: 1,
    basis: "auto"
};
exports.default = FlexItem;
