// @ff/react/ManipSource.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
////////////////////////////////////////////////////////////////////////////////
class ManipSource {
    constructor(element, traits) {
        traits = traits || {};
        traits.preventDefault = traits.preventDefault === undefined ? true : traits.preventDefault;
        this.element = element;
        this.traits = traits;
        this.onMouseDown = this.onMouseDown.bind(this);
        this.onMouseMove = this.onMouseMove.bind(this);
        this.onMouseUp = this.onMouseUp.bind(this);
        this.onMouseWheel = this.onMouseWheel.bind(this);
        this.onTouchStart = this.onTouchStart.bind(this);
        this.onTouchMove = this.onTouchMove.bind(this);
        this.onTouchEnd = this.onTouchEnd.bind(this);
        this.onTouchCancel = this.onTouchCancel.bind(this);
        this.onContextMenu = this.onContextMenu.bind(this);
        this.addEventListeners(element, false);
    }
    detach() {
        this.stopCapture();
        this.removeEventListeners(this.element, false);
    }
    setListener(listener) {
        this.listener = listener;
    }
    startCapture() {
        if (!this.isCaptured) {
            this.removeEventListeners(this.element, false);
            this.addEventListeners(document, true);
            this.isCaptured = true;
        }
    }
    stopCapture() {
        if (this.isCaptured) {
            this.removeEventListeners(document, true);
            this.addEventListeners(this.element, false);
        }
        this.isCaptured = false;
    }
    onManipBegin(event) {
        //console.log("onManipBegin", event);
        if (this.listener) {
            return this.listener.onManipBegin(event);
        }
        return true;
    }
    onManipUpdate(event) {
        //console.log("onManipUpdate", event);
        if (this.listener) {
            this.listener.onManipUpdate(event);
        }
    }
    onManipEnd(event) {
        //console.log("onManipEnd", event);
        if (this.listener) {
            this.listener.onManipEnd(event);
        }
    }
    onManipEvent(event) {
        //console.log("onManipEvent", event);
        if (this.listener) {
            this.listener.onManipEvent(event);
        }
    }
    onMouseDown(event) {
        this.setPropagation(event);
        const manipEvent = this.manipFromMouseEvent(event, "down");
        if (this.isActive) {
            this.onManipUpdate(manipEvent);
        }
        else {
            this.isActive = this.onManipBegin(manipEvent);
            if (this.isActive) {
                this.activationEvent = manipEvent;
                if (this.traits.draggable) {
                    this.startCapture();
                }
            }
        }
        this.lastEvent = manipEvent;
    }
    onMouseMove(event) {
        this.setPropagation(event);
        const manipEvent = this.manipFromMouseEvent(event, "move");
        this.onManipUpdate(manipEvent);
        this.lastEvent = manipEvent;
    }
    onMouseUp(event) {
        this.setPropagation(event);
        const manipEvent = this.manipFromMouseEvent(event, "up");
        if (this.isActive && manipEvent.buttons == 0) {
            this.onManipEnd(manipEvent);
            this.isActive = false;
            this.stopCapture();
            this.activationEvent = null;
        }
        else {
            this.onManipUpdate(manipEvent);
        }
        this.lastEvent = manipEvent;
    }
    onMouseWheel(event) {
        this.setPropagation(event);
        const manipEvent = this.manipFromMouseEvent(event, "wheel");
        this.onManipEvent(manipEvent);
    }
    onTouchStart(event) {
        this.setPropagation(event);
        event.preventDefault(); // prevent generation of additional mouse events
        const manipEvent = this.manipFromTouchEvent(event, "down");
        if (this.isActive) {
            this.onManipUpdate(manipEvent);
        }
        else {
            this.isActive = this.onManipBegin(manipEvent);
            if (this.isActive) {
                this.activationEvent = manipEvent;
                if (this.traits.draggable) {
                    this.startCapture();
                }
            }
        }
        this.lastEvent = manipEvent;
    }
    onTouchMove(event) {
        this.setPropagation(event);
        event.preventDefault(); // prevent generation of additional mouse events
        const manipEvent = this.manipFromTouchEvent(event, "move");
        this.onManipUpdate(manipEvent);
        this.lastEvent = manipEvent;
    }
    onTouchEnd(event) {
        this.setPropagation(event);
        event.preventDefault(); // prevent generation of additional mouse events
        const manipEvent = this.manipFromTouchEvent(event, "up");
        if (this.isActive && manipEvent.touches.length == 0) {
            this.onManipEnd(manipEvent);
            this.isActive = false;
            this.stopCapture();
            this.pinchDistance = 0;
            this.activationEvent = null;
            this.lastEvent = null;
        }
        else {
            this.onManipUpdate(manipEvent);
        }
        this.lastEvent = manipEvent;
    }
    onTouchCancel(event) {
        this.setPropagation(event);
        event.preventDefault(); // prevent generation of additional mouse events
        const manipEvent = this.manipFromTouchEvent(event, "cancel");
        if (this.isActive && manipEvent.touches.length == 0) {
            this.onManipEnd(manipEvent);
            this.isActive = false;
            this.stopCapture();
            this.pinchDistance = 0;
            this.activationEvent = null;
            this.lastEvent = null;
        }
        else {
            this.onManipUpdate(manipEvent);
        }
        this.lastEvent = manipEvent;
    }
    onContextMenu(event) {
        this.setPropagation(event);
    }
    setPropagation(event) {
        if (this.traits.stopPropagation === true) {
            event.stopPropagation();
        }
        if (this.traits.preventDefault === true) {
            event.preventDefault();
        }
    }
    manipFromMouseEvent(event, type) {
        let wheelEvent = event;
        return {
            source: "mouse",
            target: event.target,
            currentTarget: event.currentTarget,
            manip: this,
            type: type,
            isActive: this.isActive,
            screenX: event.screenX,
            screenY: event.screenY,
            clientX: event.clientX,
            clientY: event.clientY,
            offsetX: event.offsetX,
            offsetY: event.offsetY,
            movementX: event.movementX || (this.lastEvent ? event.screenX - this.lastEvent.screenX : 0),
            movementY: event.movementY || (this.lastEvent ? event.screenY - this.lastEvent.screenY : 0),
            centerX: event.screenX,
            centerY: event.screenY,
            button: event.button,
            buttons: event.buttons,
            wheel: type === "wheel" ? wheelEvent.deltaY : 0,
            changedTouches: null,
            targetTouches: null,
            touches: null,
            pinchDistance: 0,
            pinchFactor: 0,
            shiftKey: event.shiftKey,
            ctrlKey: event.ctrlKey,
            altKey: event.altKey,
            metaKey: event.metaKey
        };
    }
    manipFromTouchEvent(event, type) {
        const manipEvent = {
            source: "touch",
            target: event.target,
            currentTarget: event.currentTarget,
            manip: this,
            type: type,
            isActive: this.isActive,
            screenX: 0,
            screenY: 0,
            clientX: 0,
            clientY: 0,
            offsetX: 0,
            offsetY: 0,
            movementX: 0,
            movementY: 0,
            centerX: 0,
            centerY: 0,
            button: 0,
            buttons: 0,
            wheel: 0,
            changedTouches: event.changedTouches,
            targetTouches: event.targetTouches,
            touches: event.touches,
            pinchDistance: 0,
            pinchFactor: 0,
            shiftKey: event.shiftKey,
            ctrlKey: event.ctrlKey,
            altKey: event.altKey,
            metaKey: event.metaKey
        };
        const touches = event.touches;
        const count = touches.length;
        if (count > 0) {
            manipEvent.screenX = touches[0].screenX;
            manipEvent.screenY = touches[0].screenY;
            manipEvent.clientX = touches[0].clientX;
            manipEvent.clientY = touches[0].clientY;
            if (this.lastEvent) {
                manipEvent.movementX = manipEvent.screenX - this.lastEvent.screenX;
                manipEvent.movementY = manipEvent.screenY - this.lastEvent.screenY;
            }
            for (let i = 0; i < count; ++i) {
                const touch = touches[i];
                manipEvent.centerX += touch.screenX;
                manipEvent.centerY += touch.screenY;
            }
            manipEvent.centerX /= count;
            manipEvent.centerY /= count;
            if (count === 2) {
                const dx = touches[1].screenX - touches[0].screenX;
                const dy = touches[1].screenY - touches[0].screenY;
                const d = Math.sqrt(dx * dx + dy * dy);
                manipEvent.pinchDistance = d;
                if (this.pinchDistance === 0) {
                    this.pinchDistance = d;
                }
                manipEvent.pinchFactor = d / this.pinchDistance;
            }
        }
        return manipEvent;
    }
    addEventListeners(target, capture) {
        target.addEventListener("mousedown", this.onMouseDown, capture);
        target.addEventListener("mousemove", this.onMouseMove, capture);
        target.addEventListener("mouseup", this.onMouseUp, capture);
        target.addEventListener("contextmenu", this.onContextMenu, capture);
        if (this.traits.touchable) {
            target.addEventListener("touchstart", this.onTouchStart, capture);
            target.addEventListener("touchmove", this.onTouchMove, capture);
            target.addEventListener("touchend", this.onTouchEnd, capture);
            target.addEventListener("touchcancel", this.onTouchCancel, capture);
        }
        if (this.traits.scrollable) {
            target.addEventListener("wheel", this.onMouseWheel, capture);
        }
    }
    removeEventListeners(target, capture) {
        target.removeEventListener("mousedown", this.onMouseDown, capture);
        target.removeEventListener("mousemove", this.onMouseMove, capture);
        target.removeEventListener("mouseup", this.onMouseUp, capture);
        target.removeEventListener("contextmenu", this.onContextMenu, capture);
        if (this.traits.touchable) {
            target.removeEventListener("touchstart", this.onTouchStart, capture);
            target.removeEventListener("touchmove", this.onTouchMove, capture);
            target.removeEventListener("touchend", this.onTouchEnd, capture);
            target.removeEventListener("touchcancel", this.onTouchCancel, capture);
        }
        if (this.traits.scrollable) {
            target.removeEventListener("wheel", this.onMouseWheel, capture);
        }
    }
}
exports.default = ManipSource;
