// @ff/react/ManipDispatcher.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
////////////////////////////////////////////////////////////////////////////////
class ManipDispatcher {
    constructor() {
        this.handlers = [];
        this.activeHandler = null;
    }
    addManip(handler) {
        this.handlers.push(handler);
    }
    insertManip(index, handler) {
        this.handlers.splice(index, 0, handler);
    }
    removeManip(handler) {
        const index = this.handlers.indexOf(handler);
        if (index >= 0) {
            this.handlers.splice(index, 1);
        }
    }
    onPointer(event) {
        let consumed = false;
        if (this.activeHandler) {
            consumed = this.activeHandler.onPointer(event);
            if (!consumed) {
                this.activeHandler = null;
            }
        }
        else {
            for (let i = 0; i < this.handlers.length; ++i) {
                const handler = this.handlers[i];
                consumed = handler.onPointer(event);
                if (consumed) {
                    this.activeHandler = handler;
                    this.activeHandlerIndex = i;
                    break;
                }
            }
        }
        return consumed;
    }
    onTrigger(event) {
        let consumed = false;
        if (this.activeHandler) {
            consumed = this.activeHandler.onTrigger(event);
        }
        else {
            for (let i = 0; i < this.handlers.length; ++i) {
                const handler = this.handlers[i];
                consumed = handler.onTrigger(event);
                if (consumed) {
                    break;
                }
            }
        }
        return consumed;
    }
}
exports.default = ManipDispatcher;
