// @ff/react/ManipTarget.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
const ManipTarget_1 = require("@ff/browser/ManipTarget");
class ManipTarget extends React.Component {
    constructor(props) {
        super(props);
        this.elementRef = React.createRef();
        this.manipHelper = new ManipTarget_1.default();
        this.manipHelper.next = this;
    }
    render() {
        const { className, children } = this.props;
        const manipHelper = this.manipHelper;
        const props = {
            className,
            ref: this.elementRef,
            "touch-action": "none",
            onPointerDown: manipHelper.onPointerDown,
            onPointerMove: manipHelper.onPointerMove,
            onPointerUp: manipHelper.onPointerUpOrCancel,
            onPointerCancel: manipHelper.onPointerUpOrCancel,
            onDoubleClick: manipHelper.onDoubleClick,
            onContextMenu: manipHelper.onContextMenu,
            onWheel: manipHelper.onWheel
        };
        return React.createElement("div", props, children);
    }
    onPointer(event) {
        const props = this.props;
        if (props.handler) {
            return props.handler.onPointer(event);
        }
        if (props.onPointer) {
            return props.onPointer(event);
        }
    }
    onTrigger(event) {
        const props = this.props;
        if (props.handler) {
            return props.handler.onTrigger(event);
        }
        if (props.onTrigger) {
            return props.onTrigger(event);
        }
    }
}
exports.default = ManipTarget;
ManipTarget.defaultProps = {
    className: "ff-manip-target"
};
