// @ff/react/Overlay.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
const ReactDOM = require("react-dom");
class Overlay extends React.Component {
    constructor(props) {
        super(props);
        this.element = document.createElement("div");
    }
    componentDidMount() {
        document.body.appendChild(this.element);
    }
    componentWillUnmount() {
        document.body.removeChild(this.element);
    }
    render() {
        const { className, style, zIndex, children } = this.props;
        const styles = Object.assign({}, Overlay.style, style);
        if (zIndex !== undefined) {
            styles.zIndex = zIndex;
        }
        return ReactDOM.createPortal(React.createElement("div", { className: className, style: styles }, children), this.element);
    }
}
exports.default = Overlay;
Overlay.defaultProps = {
    className: "ff-overlay",
    zIndex: 1000
};
Overlay.style = {
    position: "absolute",
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
};
