// @ff/react/TabContent.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
const FlexItem_1 = require("./FlexItem");
class TabContentContainer extends React.Component {
    render() {
        const { className, style, children } = this.props;
        return (React.createElement(FlexItem_1.default, { className: className, style: style }, children));
    }
}
exports.TabContentContainer = TabContentContainer;
TabContentContainer.defaultProps = {
    className: "ff-tab-content-container"
};
class TabContentItem extends React.Component {
    render() {
        const { className, style, active, children } = this.props;
        const styles = Object.assign({}, TabContentItem.style, style, { display: active ? "initial" : "none" });
        return (React.createElement("div", { className: className, style: styles }, children));
    }
}
exports.TabContentItem = TabContentItem;
TabContentItem.defaultProps = {
    className: "ff-tab-content-item"
};
TabContentItem.style = {
    position: "absolute",
    left: 0, top: 0, right: 0, bottom: 0,
    display: "initial",
    overflow: "hidden"
};
