// @ff/react/TitleBar.js
"use strict";
/**
 * FF Typescript Foundation Library
 * Copyright 2018 Ralph Wiedemeier, Frame Factory GmbH
 *
 * License: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const React = require("react");
class TitleBar extends React.Component {
    constructor(props) {
        super(props);
    }
    render() {
        const { className, text, children, } = this.props;
        let barStyle = {};
        let textStyle = {
            whiteSpace: "nowrap",
            overflow: "hidden",
            textOverflow: "ellipsis"
        };
        let buttonsStyle = {
            float: "right",
            whiteSpace: "nowrap",
            overflow: "hidden"
        };
        return (React.createElement("div", { className: className, style: barStyle },
            React.createElement("div", { className: "ff-buttons", style: buttonsStyle }, this.props.children),
            React.createElement("div", { className: "ff-text", style: textStyle }, this.props.text)));
    }
}
exports.default = TitleBar;
TitleBar.defaultProps = {
    className: "ff-title-bar",
    text: "Title Bar"
};
