// common/types.js
"use strict";
/**
 * 3D Foundation Project
 * Copyright 2019 Smithsonian Institution
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var MachineState;
(function (MachineState) {
    MachineState[MachineState["Idle"] = 0] = "Idle";
    MachineState[MachineState["Busy"] = 1] = "Busy";
    MachineState[MachineState["Disabled"] = 2] = "Disabled";
    MachineState[MachineState["Unavailable"] = 3] = "Unavailable";
    MachineState[MachineState["Error"] = 4] = "Error";
})(MachineState = exports.MachineState || (exports.MachineState = {}));
